rpc: re-index missing events

JayT106 opened the pull request:
Implemented an unsafe API `re_index` to the RPC server.
The node operator need to enable this API by setting the unsafe flag in `config.toml`

Note: the eventsink (PSQL sink) will not remove the duplicate records if there are existing events in the given re-index range.


Closes: #6519 , #4371

tychoish commented:
> Note: the eventsink (PSQL sink) will not remove the duplicate records if there are existing events in the given resync range.

Would it be possible to do these as upserts to avoid this limitation?

JayT106 commented:
> > Note: the eventsink (PSQL sink) will not remove the duplicate records if there are existing events in the given resync range.
> 
> Would it be possible to do these as upserts to avoid this limitation?

do you mean the user can give the range like `1-10, 12, `

JayT106 commented:
> > Note: the eventsink (PSQL sink) will not remove the duplicate records if there are existing events in the given resync range.
> 
> Would it be possible to do these as upserts to avoid this limitation?

Do you mean the operator can give the range and can skip some blocks?

Or you want the SQL server can check the duplicate data we the indexer is inserting the records because the core is not able to check the database record under the current implementation.

marbar3778 commented on rpc/core/routes.go:
```diff
@@ -58,4 +58,5 @@ func (env *Environment) AddUnsafe(routes RoutesMap) {
 	routes["dial_seeds"] = rpc.NewRPCFunc(env.UnsafeDialSeeds, "seeds", false)
 	routes["dial_peers"] = rpc.NewRPCFunc(env.UnsafeDialPeers, "peers,persistent,unconditional,private", false)
 	routes["unsafe_flush_mempool"] = rpc.NewRPCFunc(env.UnsafeFlushMempool, "", false)
+	routes["resync"] = rpc.NewRPCFunc(env.UnsafeResync, "start_height,end_height", false)
```
reindex or something more related to events may be a better name. This could be confused with resyncing blocks from the network.

JayT106 replied:
Sure, I'll change to `reindex` and the relate docs.

tychoish commented:
my understanding is that many database have an "insert if <value for key in a unique index> doesn't exist, and otherwise update" operation that they call "upsert" which might be right for this kind of "reindex" operation (it might also be slow at the kind of scale we're talking, but that might be ok, particularly if it's a background operation and you can prevent more than one-reindexing operation from happening at once)

JayT106 commented:
> my understanding is that many database have an "insert if doesn't exist, and otherwise update" operation that they call "upsert" which might be right for this kind of "reindex" operation (it might also be slow at the kind of scale we're talking, but that might be ok, particularly if it's a background operation and you can prevent more than one-reindexing operation from happening at once)

Ah, I see, thanks! It will require the EventSink to implement one more function to do the upsert operation.
@alexanderbez what do you think to add the `upsertEvents` function to the EventSink interface, Or revising the `IndexTxEvents` and `IndexBlockEvents` to use the `upsert` operation?

alexanderbez commented:
We should not be upserting at all. We should only be inserting records if they do not exist. There is no need to update the interface(s). Instead, the psql sink can do this logic for us in the `IndexTxEvents` and `IndexBlockEvents` methods.

JayT106 commented:
Added another PR for the upsert operation in the PSQL eventsink.
#6556

JayT106 commented on rpc/core/dev.go:
```diff
@@ -1,12 +1,104 @@
+		return nil, fmt.Errorf("no event sink has been enabled")
+	}
+
+	for i := start; i <= end; i++ {
```
good point, thanks!

JayT106 commented on rpc/core/dev.go:
```diff
@@ -1,12 +1,104 @@
 }
+
+// UnsafeReIndex re-index the block/transaction events into the eventsinks.
+func (env *Environment) UnsafeReIndex(
```
But the indexer doesn't know anything about the blocks and states.

Yes, the re-index call will interleave with the regular indexer operations at the same time. Inside the Index functions, the kv/psql DB process the operations atomically. And there is no sharing instance (except the DB) in the Index functions so I think it would be okay?

marbar3778 reviewed (approved):
utACK

alexanderbez reviewed (commented):
There was one thing that I'm not sure was discussed with this PR or the original intent. Re-indexing could potentially take a LONG time, say if the Postgres instance is slow and/or the amount of blocks to re-index is large. In this case, it's not safe to execute as an RPC call, as the RPC call will timeout and you have no idea of the state of the re-indexing.

It seems to be that re-indexing should be stateful and a curser or ID can be given back to the caller to check the state of the current re-indexing.

JayT106 commented:
> There was one thing that I'm not sure was discussed with this PR or the original intent. Re-indexing could potentially take a LONG time, say if the Postgres instance is slow and/or the amount of blocks to re-index is large. In this case, it's not safe to execute as an RPC call, as the RPC call will timeout and you have no idea of the state of the re-indexing.
> 
> It seems to be that re-indexing should be stateful and a curser or ID can be given back to the caller to check the state of the current re-indexing.

@alexanderbez  PR seems to be merged, do you want to revert it or let me add another PR for addressing your concerns?

alexanderbez commented:
I'm not totally sure tbh. I feel like we should revert it and think about it more. @marbar3778 what are your thoughts?

marbar3778 commented:
> I'm not totally sure tbh. I feel like we should revert it and think about it more. @marbar3778 what are your thoughts?

I think returning saying we have started the reindexing service is fine. We dont have to wait to respond since they will see things being indexed via the eventsink or subscribe endpoint

alexanderbez commented:
This would require additional work:

- Requiring the request to be stateful
- Requiring the ability for clients to check the status
